components/uniset/src/builder.rs

```diff
@@ -0,0 +1,755 @@
+        assert_eq!(builder.intervals, check);
+    }
+
+    #[test]
+    fn test_add_to_empty() {
+        let mut builder = UnicodeSetBuilder::new();
+        builder.add(0, 10);
+        assert_eq!(builder.intervals, vec![0, 10]);
+    }
+
+    #[test]
+    fn test_add_invalid() {
+        let mut builder = UnicodeSetBuilder::new();
+        builder.add(0, 0);
+        builder.add(5, 0);
+        assert_eq!(builder.intervals, vec![]);
+    }
+
+    #[test]
+    fn test_add_to_start() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(0, 5);
+        let expected = vec![0, 5, 10, 20, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_start_overlap() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(0, 15);
+        let expected = vec![0, 20, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_end() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(60, 70);
+        let expected = vec![10, 20, 40, 50, 60, 70];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_end_overlap() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(45, 70);
+        let expected = vec![10, 20, 40, 70];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_middle_no_overlap() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(25, 27);
+        let expected = vec![10, 20, 25, 27, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_middle_inside() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(10, 20);
+        let expected = vec![10, 20, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_middle_left_overlap() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(15, 25);
+        let expected = vec![10, 25, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_middle_right_overlap() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(30, 40);
+        let expected = vec![10, 20, 30, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_full_encompass() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(0, 60);
+        let expected = vec![0, 60];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_add_to_partial_encompass() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.add(0, 35);
+        let expected = vec![0, 35, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
```

Comment:
You should test your functions, especially `add` and `remove`, with ranges that bump up against an existing range, to test the behavior about whether the binary search is `Ok` or `Err`.  For example:

```rust
    #[test]
    fn test_add_aligned_front() {
        let mut builder = generate_tester(vec![10, 20, 40, 50]);
        builder.add(5, 10);
        let expected = vec![5, 20, 40, 50];
        assert_eq!(builder.intervals, expected);
    }
```

Also with ranges such as:

- 10, 15
- 15, 20
- 20, 25
- 35, 40
- 20, 40
- 50, 55

Reply:
agreed and added

Reply:
Would it be better if I included source comments?

I have written those tests specifically and ready for commit, but since `add` and `remove` rely on non-empty ranges (`start < end`), `builder.add(x,  x)` is not possible, and instead would look like `builder.add(x, x + 1)` which falls into the same behavior as the adjacent cases as a range.

Reply:
Yeah, source comments would help since there is an assumption that is not stated.